HeuristicNew.java

```diff
@@ -0,0 +1,256 @@
+
+        // System.out.print("Col heights: ");
+
+        for (int col = 0; col < field[0].length; col++) {
+            int row = field.length - 1;
+
+            while (row >= 0 && field[row][col] == 0) {
+                row--;
+            }
+            colHeights[col] = row + 1;
+            // System.out.print(colHeights[col] + " ");
+        }
+        // System.out.println();
+
+        return colHeights;
+    }
+
+    // Feature 5
+    // Returns weighted number of holes
+    private static double wNumHoles(int[][] field, int[] colHeights) {
+        int numHoles = 0;
+
+        for (int col = 0; col < colHeights.length; col++) {
+            for (int row = colHeights[col] - 1; row >= 0; row--) {
+                if (field[row][col] == 0) {
+                    numHoles++;
+                }
+            }
+        }
+
+        // System.out.println("Number of holes = " + numHoles);
+
+        return getNextWeight() * numHoles;
+    }
+
+    // Feature 6
+    private static double weightedWellSums(int[][] field) {
+        return getNextWeight() * numColTransitions(field);
+    }
+
+    // A well is a succession of empty cells such that their left cells and right cells are both filled.
+    // Example:
+    // 1101100010
+    // 1101101010
+    // 1101101000
+    // returns 8
+    private static int wellSums(int[][] field) {
+        int count = 0;
+        int numRows = field.length;
+        int numCols = field[0].length;
+        for (int i = 0; i < numRows; i++) {
+            for (int j = 0; j < numCols; j++) {
+                if (field[i][j] == 0){ // if cell is empty
+                        if (j > 0 && j < numCols - 1 && field[i][j - 1] != 0 && field[i][j + 1] != 0) {
+                            count += exploreWell(i, j, field);
+                        } else if (j == 0 && field[i][j+1] !=0) {
+                            count += exploreWellLeftSide(i, field);
+                        } else if (j == numCols - 1 && field[i][j-1] != 0){
+                            count += exploreWellRightSide(i, field);
+                        }
+                }
+
+            }
+        }
+        //System.out.println("Overall Well sum: " + count);
+        return count;
+    }
+
+    // explore up the rows within the column
+    private static int exploreWell(int startRow, int startColumn, int[][] field) {
+	    // if this empty column is part of an existing well. Don't bother
+	    if (startRow != 0
+                && field[startRow-1][startColumn] == 0
+                && field[startRow-1][startColumn - 1] != 0
+                && field[startRow-1][startColumn + 1] != 0) {
+            return 0;
+        }
+
+	    int count = 0;
+	    while(field[startRow][startColumn] == 0
+                && field[startRow][startColumn-1] != 0
+                && field[startRow][startColumn+1] != 0) {
+	        count++;
+	        startRow++;
+        }
+        if (field[startRow][startColumn] != 0) {
```

Comment:
Why would this case every happen?

Reply:
`field[startRow][startColumn] == 0 && field[startRow][startColumn-1] != 0 && field[startRow][startColumn+1] != 0` can evaluate false when the 'walls' cells i.e. `field[startRow][startColumn-1]` and  `field[startRow][startColumn+1]`  are empty, but it is possible that `field[startRow][startColumn] != 0`.

T AND F AND F evaluates to F.

Simply put, you think you've explored a well, but actually, you've explored a hole.

010
101
101
101